[wolfsmartset] Initial contribution

BoBiene opened the pull request:
New Binding for Wolf Smartset (https://www.wolf-smartset.com/)

daMihe commented on bundles/org.openhab.binding.wolfsmartset/README.md:
```diff
@@ -0,0 +1,45 @@
+# Wolf Smartset Binding
```
The README has to describe how to use the binding with the device. This isn't here the case. It's also mostly in german which is not readable for a great part of the community. Please rewrite it.

BoBiene replied:
correct, the readme is currently just a placeholder.
I wanted to get some initial thoughts on my code to refine it.

daMihe commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/WolfSmartsetBindingConstants.java:
```diff
@@ -0,0 +1,92 @@
+ * The {@link WolfSmartsetBindingConstants} class defines common constants that are
+ * used across the whole binding.
+ *
+ * @author Bo Biene - Initial contribution
```
Just mentioning as your username is the same as the name here: Please use your real name as described in the [Contribution guidelines](https://www.openhab.org/docs/developer/contributing.html).

BoBiene replied:
My Username is my real name 😉

daMihe commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/api/WolfSmartsetApi.java:
```diff
@@ -0,0 +1,650 @@
+            .collect(StringBuilder::new, StringBuilder::appendCodePoint, StringBuilder::append).toString());
+    private static final String USERAGENT = "Android-7.1.1-1.0.0-ONEPLUS A3010-136-" + AGENT_ID
+            + " APP/wolf-smartset.smarthome APPV/62830";
+    // private static Locale locale = Locale.getDefault();
```
Please remove non describing comments.

daMihe commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/api/WolfSmartsetApi.java:
```diff
@@ -0,0 +1,650 @@
+    private static final String UNEXPECTED = "Unexpected :";
+    private static final String AGENT_ID = (new Random().ints(65, 70).limit(13)
+            .collect(StringBuilder::new, StringBuilder::appendCodePoint, StringBuilder::append).toString());
+    private static final String USERAGENT = "Android-7.1.1-1.0.0-ONEPLUS A3010-136-" + AGENT_ID
```
Is this User-Agent required for work?

daMihe commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/api/WolfSmartsetApi.java:
```diff
@@ -0,0 +1,650 @@
+        setDelay(delay);
+    }
+
+    public void stopClient() {
```
All public methods in your regular code should be documented.

daMihe commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/api/WolfSmartsetApi.java:
```diff
@@ -0,0 +1,650 @@
+        return "";
+    }
+
+    public synchronized boolean login() {
```
please do not mix public and private methods for better readability.

daMihe commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetAccountBridgeHandler.java:
```diff
@@ -0,0 +1,279 @@
+            api = new WolfSmartsetApi(username, password, httpClient, scheduler);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_PENDING, "Checking authorization");
+            scheduleRefreshJob();
+        } catch (WolfSmartsetCloudException e) {
```
Please try to be more specific (login data wrong/missing connection/etc.) and do not force the user to interpret stacktraces, wherever possible. Stacktraces are usually better in deeper levels so a user with more knowledge can read them if needed. But a normal user is not able to read them.

daMihe commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetAccountBridgeHandler.java:
```diff
@@ -0,0 +1,279 @@
+                            }
+                        }
+
+                    } else {
```
Please remove unnecessary blocks.

daMihe commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetSystemBridgeHandler.java:
```diff
@@ -0,0 +1,250 @@
+        savedUnits = listUnits;
+    }
+
+    public void updateSystemState(@Nullable GetSystemStateListDTO systemState) {
```
useless?

BoBiene replied:
i have implemented `updateSystemState`

daMihe reviewed (changes requested):
The review was now quite quick and not in depth. Please fix the issues generally, not only where i've added comments. I'll do a more deep review after, not before next week.

BoBiene commented:
> The review was now quite quick and not in depth. Please fix the issues generally, not only where i've added comments. I'll do a more deep review after, not before next week.

@daMihe i hope i have fixed all issues.

daMihe commented:
I need more time getting my IDE work again...

Never the less, your commit messages are not in common format, they should be formatted like "[wolfsmartset] something short, but useful". Also you are using a pseudonymous address for signing off commits which is not allowed according to the contribution document:

> using your real name (sorry, no pseudonyms or anonymous contributions.) and an e-mail address under which you can be reached (sorry, no github noreply e-mail addresses (such as username@users.noreply.github.com) or other non-reachable addresses are allowed).

Please get familiar with `git rebase -i` and correct this (please do not reopen another pull request, just take a offline backup and push only after `git log` looks good). Signing off is just a line at the end of the extended commit message, so you are not breaking anything correcting this. Signing off a commit you agree to the contribution guide. Please read it carefully.

BoBiene commented:
> I need more time getting my IDE work again...
> 
> Never the less, your commit messages are not in common format, they should be formatted like "\[wolfsmartset\] something short, but useful". Also you are using a pseudonymous address for signing off commits which is not allowed according to the contribution document:
> 
> using your real name (sorry, no pseudonyms or anonymous contributions.) and an e-mail address under which you can be reached (sorry, no github noreply e-mail addresses (such as [username@users.noreply.github.com](mailto:username@users.noreply.github.com)) or other non-reachable addresses are allowed).
> 
> Please get familiar with `git rebase -i` and correct this (please do not reopen another pull request, just take a offline backup and push only after `git log` looks good). Signing off is just a line at the end of the extended commit message, so you are not breaking anything correcting this. Signing off a commit you agree to the contribution guide. Please read it carefully.

Hi @daMihe the history should now look correctly

BoBiene commented:
Build Failed but as far as I can see it is because of the digitalstrom binding. What can I do to solve this?

kaikreuzer commented:
Have it rebuilt. (running...)

BoBiene commented:
/rebuild

Build faild with timeout:  
\`\`\`\`

Build timed out (after 240 minutes). Marking the build as aborted. ERROR: Processing failed due to a bug in the code. Please report this to the issue tracker ([https://jenkins.io/redirect/report-an-issue](https://jenkins.io/redirect/report-an-issue)). java.lang.NullPointerException at hudson.maven.AbstractMavenBuilder.end(AbstractMavenBuilder.java:101) at hudson.maven.MavenModuleSetBuild$MavenModuleSetBuildExecution.doRun(MavenModuleSetBuild.java:883) at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:513) at hudson.model.Run.execute(Run.java:1907) at hudson.maven.MavenModuleSetBuild.run(MavenModuleSetBuild.java:543) at hudson.model.ResourceController.execute(ResourceController.java:97) at hudson.model.Executor.run(Executor.java:429)\`\`\`\`

kaikreuzer commented:
This time it succeeded. 👍

BoBiene commented:
> This time it succeeded. 👍 

Thanks 🙂

daMihe commented on bundles/org.openhab.binding.wolfsmartset/README.md:
```diff
@@ -0,0 +1,96 @@
+
+| Heating system                            | WOLF Link home        | WOLF Link pro      |
+|-------------------------------------------|-----------------------|--------------------|
+| Gas condensing boiler CGB-2, CGW-2, CGS-2 | :heavy_check_mark: | :heavy_check_mark: |
```
I'm not sure whether emoticons work this way. In doubt i would use unicode emoticons.

daMihe commented on bundles/org.openhab.binding.wolfsmartset/README.md:
````diff
@@ -0,0 +1,96 @@
+:black_square_button: possible in connection with a WOLF Link home compatible heater,
+full functionality only for devices with current software version.
+
+``` * ``Modbus interface required in the device,
````
is the formatting here behaving as expected?

daMihe commented on bundles/org.openhab.binding.wolfsmartset/README.md:
````diff
@@ -0,0 +1,96 @@
+
+``` * ``Modbus interface required in the device,
+Special programming cannot be mapped.
+
````
Please provide an textual configuration sample, also.

BoBiene replied:
This binding should only be used with discovery and not with manual configuration. I would prefer to not document this.

And as far as i understood, the new preferred way to configure openHAB, is via the UI and not via the textual configuration.

daMihe replied:
At least it should be documented as such. Please provide a short information.

daMihe reviewed (changes requested):
Done a more deep review. This is only what i can see here and in eclipse, i'm not able to do a functional test. I'm not able to check English texts as i'm not a native speaker. Please check the output of the SAT-plugin (see output and created file during `mvn install -pl :org.openhab.binding.wolfsmartset`, there are some further hints.

daMihe commented:
@kaikreuzer is this now "cre"?

BoBiene commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/api/WolfSmartsetApi.java:
```diff
@@ -0,0 +1,622 @@
+        return parameterValues;
+    }
+
+    private void startClient() throws WolfSmartsetCloudException {
+        if (!httpClient.isStarted()) {
+            try {
+                httpClient.start();
+            } catch (Exception e) {
+                throw new WolfSmartsetCloudException("No http client cannot be started: " + e.getMessage(), e);
+            }
+        }
+        setDelay(delay);
+    }
+
+    public void stopClient() {
+        try {
+            stopProcessJob();
+            requestQueue.forEach(queueEntry -> queueEntry.future.completeExceptionally(new CancellationException()));
+            this.httpClient.stop();
+        } catch (Exception e) {
+            logger.debug("Error stopping httpclient :{}", e.getMessage(), e);
+        }
+    }
```
@daMihe i copied a lot from the miio binding, and this comes from [https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/cloud/MiCloudConnector.java#L119](https://github.com/openhab/openhab-addons/blob/main/bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/cloud/MiCloudConnector.java#L119) 

@marcelrv do you know the reason for the httpClient.stop?

daMihe replied:
@marcelrv uses in miio-binding a self maintained httpclient and creates it [here](https://github.com/openhab/openhab-addons/blob/2cc08ba8bd28ae1647bf8e3e2637cb978b2cf3bb/bundles/org.openhab.binding.miio/src/main/java/org/openhab/binding/miio/internal/cloud/CloudConnector.java#L94). Whether that is useful there is out of scope here. If you do not need to modify the `httpclient` itself, you can and should use the shared instance. But then you must not modify it.

marcelrv replied:
Indeed as @daMihe indicated, it is because I don't use the shared instance.
Normally this should not be required when using the (recommended way) shared httpclient

BoBiene commented:
> Done a more deep review. This is only what i can see here and in eclipse, i'm not able to do a functional test. I'm not able to check English texts as i'm not a native speaker. Please check the output of the SAT-plugin (see output and created file during `mvn install -pl :org.openhab.binding.wolfsmartset`, there are some further hints.

@daMihe to do a functional test, you can create an account for [https://www.wolf-smartset.com/,](https://www.wolf-smartset.com/,) send me the mail-address used and i can share my heating device with your account.

BoBiene commented:
@daMihe i have pushed the requested changes.

BoBiene commented:
hmm, build failed. is this due to the release of 3.1? How can i fix this issue?

daMihe commented:
Please update your pom version to `3.2.0-SNAPSHOT`

daMihe commented:
> @daMihe i have pushed the requested changes.

You forgot some. Did you may ignore the "19 hidden conversations"?

daMihe commented:
> > [...]
> 
> @daMihe to do a functional test, you can create an account for https://www.wolf-smartset.com/, send me the mail-address used and i can share my heating device with your account.

I can do but i don't require this (i also got no feedback about functionality in my binding - and had to correct it after merge a bit). I assume you've tried your binding and checked the values for plausibility (which is what i can't do next) so i'll trust you in that point without check. It must not interfere with other plugins especially while building.

BoBiene commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/api/WolfSmartsetApi.java:
```diff
@@ -0,0 +1,622 @@
+            Request request = httpClient.POST(url).timeout(REQUEST_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+            request.header(HttpHeader.CONTENT_TYPE, "application/x-www-form-urlencoded");
+
+            var encodedUser = URLEncoder.encode(username, StandardCharsets.UTF_8);
+            var encodedPassword = URLEncoder.encode(password, StandardCharsets.UTF_8);
+            var authRequestBody = "grant_type=password&username=" + encodedUser + "&password=" + encodedPassword;
```
I tried, but the wolf backend is very picky with the parameter format. So i did it this way

daMihe replied:
Please add a comment documenting this.

BoBiene commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/api/WolfSmartsetApi.java:
```diff
@@ -0,0 +1,622 @@
+
+            request.header(HttpHeader.AUTHORIZATION, serviceToken);
+            request.method(HttpMethod.GET);
+            request.header(HttpHeader.CONTENT_TYPE, "application/x-www-form-urlencoded");
+
+            for (Entry<String, String> entry : params.entrySet()) {
+                logger.debug("Send request param: {}={} to {}", entry.getKey(), entry.getValue().toString(), url);
+                request.param(entry.getKey(), entry.getValue());
+            }
```
I must deal with the format the Wolf-Smartset APP uses. I had to reverse-engenier the http requests done by the Application

BoBiene commented:
> @daMihe i have pushed the requested changes.
> 
> You forgot some. Did you may ignore the "19 hidden conversations"?

Yes, you right. Just pushed them

BoBiene reviewed (commented):
I have completed all requested changes

daMihe commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/api/WolfSmartsetApi.java:
```diff
@@ -0,0 +1,617 @@
+            if (response.getStatus() == HttpStatus.NOT_FOUND_404) {
+                throw new WolfSmartsetCloudException("Invalid request, not found " + request.getURI());
+            } else if (response.getStatus() == HttpStatus.TOO_MANY_REQUESTS_429) {
+                Thread.sleep(30000);
```
Is this the right place to sleep? Shouldn't be this done in the next request?

daMihe commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/discovery/WolfSmartsetAccountDiscoveryService.java:
```diff
@@ -0,0 +1,153 @@
+
+    private synchronized void discoverSystems() {
+        logger.debug("WolfSmartsetDiscovery: Discovering systems");
+        var registeredSytems = bridgeHandler.getRegisteredSystems();
+        if (registeredSytems != null) {
+            for (GetSystemListDTO system : bridgeHandler.getRegisteredSystems()) {
```
please use your checked variable.

daMihe commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/discovery/WolfSmartsetSystemDiscoveryService.java:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    private void backgroundDiscover() {
+        if (!bridgeHandler.getAccountBridgeHandler().isBackgroundDiscoveryEnabled()) {
```
Nullable check fails.

daMihe reviewed (changes requested):
Some things left.

BoBiene commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/api/WolfSmartsetApi.java:
```diff
@@ -0,0 +1,617 @@
+        return parameterValues;
+    }
+
+    private void startClient() throws WolfSmartsetCloudException {
+        if (!httpClient.isStarted()) {
+            try {
+                httpClient.start();
+            } catch (Exception e) {
+                throw new WolfSmartsetCloudException("No http client cannot be started: " + e.getMessage(), e);
+            }
+        }
+        setDelay(delay);
+    }
```
@daMihe i tried to, but received exceptions that the client is not running...

daMihe replied:
Please try to isolate the issue. Do you have other bindings in your environment causing this issue? Running the binding in the demo project has no issue when commenting out the if-block. Please do not try to fix problems caused by other bindings. This makes issues harder to debug.

daMihe reviewed (changes requested):
A bit more.

BoBiene commented:
i think it need a rebuild again.

daMihe reviewed (approved):
I've triggered a rebuild. Looking into history, i'm not sure if this will help. Build times look currently broken (normally one or two hours) and "your" last build timed out (aroung 5.75 h). The build of your binding runs on my machine fine.

Please note i'm no native speaker and as such not able to check the texts in detail. I've also tested using an empty account which worked basically (for what i've seen).

BoBiene commented:
/rebuild

BoBiene commented:
@kaikreuzer / @daMihe could you please trigger a rebuild?

BoBiene commented:
@marcelrv / @kaikreuzer What else is needed for us to complete this pull request?

marcelrv commented:
I'm not one of the maintainers that can add weight on the approvers balance...
Hence suggest to request review from the add-on maintainers as you did

BoBiene commented:
@kaikreuzer can you say what is missing here?

kaikreuzer commented:
Hey @BoBiene, sorry for keeping you waiting. What is missing is mainly the time of the maintainers to review, especially now in the summer season...
I'll volunteer to do the review, but will have to see how much time I can find due to family holiday next weeks.

One wish I have after a first glance: It would be nice if you could set up the README very closely to the template that was generated, i.e. with all the sections that it listed. You are free to add appendix with further information towards the end of course. But there e.g. shouldn't be a step-by-step tutorial for the UI (screenshots age very badly!) and all parts must be described that they are valid for UI and textual configuration likewise. If it is tricky to get certain information for textual configuration, so be it, but please nonetheless list it. Users can still do a discovery once to find out device specific ids and then create a textual configuration for themselves.

mr-gosh commented:
Thanks that you still work on this @BoBiene - i look forward to use this plugin in my setup - as you know I allready use the early versions...

BoBiene commented:
@kaikreuzer ok, i did the requested changes.

BoBiene commented:
@kaikreuzer can you please trigger a rebuild?

kaikreuzer commented:
@BoBiene If you look at [the logs](https://ci.openhab.org/job/PR-openHAB-Addons/5174/console), you will see that the build failed due to this PR:
```
[ERROR] Code Analysis Tool has found 1 error(s)!
[ERROR] Please fix the errors and rerun the build.
```
Once you push a commit with a fix, the build is automatically retriggered.

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/discovery/WolfSmartsetAccountDiscoveryService.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    private void backgroundDiscover() {
+        if (!bridgeHandler.isBackgroundDiscoveryEnabled()) {
```
Wouldn't it be much better to not at all start the scheduled job at all, if background discovery is disabled?

BoBiene replied:
I've copied the code from a binding - can changes this if you are shure that it will work if the configuration changes.

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/discovery/WolfSmartsetSystemDiscoveryService.java:
```diff
@@ -0,0 +1,161 @@
+
+    private void backgroundDiscover() {
+        var accountBridgeHandler = bridgeHandler.getAccountBridgeHandler();
+        if (accountBridgeHandler == null || !accountBridgeHandler.isBackgroundDiscoveryEnabled()) {
```
See above

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetAccountBridgeHandler.java:
```diff
@@ -0,0 +1,273 @@
+            scheduleRefreshJob();
+        } catch (WolfSmartsetCloudException e) {
+            logger.error("unable to create wolf smartset api", e);
+            updateStatus(ThingStatus.UNINITIALIZED, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
```
```suggestion
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getMessage());
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetAccountBridgeHandler.java:
```diff
@@ -0,0 +1,273 @@
+        if (refreshConfigurationCounter.getAndDecrement() == 0) {
+            refreshConfigurationCounter.set(refreshIntervalConfigurationMinutes * 60);
+            if (api.login()) {
+                logger.info("AccountBridge: refreshing configuration");
```
```suggestion
                logger.debug("AccountBridge: refreshing configuration");
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetAccountBridgeHandler.java:
```diff
@@ -0,0 +1,273 @@
+        if (refreshValuesCounter.getAndDecrement() == 0) {
+            refreshValuesCounter.set(refreshIntervalValuesSeconds);
+            if (api.login()) {
+                logger.info("AccountBridge: refreshing values");
```
```suggestion
                logger.debug("AccountBridge: refreshing values");
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetSystemBridgeHandler.java:
```diff
@@ -0,0 +1,267 @@
+    /**
+     * return the associated account bridge handler
+     * 
+     * @return
```
please add a short description

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetSystemBridgeHandler.java:
```diff
@@ -0,0 +1,267 @@
+    /**
+     * return the subordinated unit handler
+     * 
+     * @return
```
please add a short description

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetSystemBridgeHandler.java:
```diff
@@ -0,0 +1,267 @@
+    /**
+     * returns the list configuration of the units available for this system
+     * 
+     * @return
```
please add a short description

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetSystemBridgeHandler.java:
```diff
@@ -0,0 +1,267 @@
+    }
+
+    /**
+     * return the associated account bridge handler
```
```suggestion
     * Return the associated account bridge handler
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetSystemBridgeHandler.java:
```diff
@@ -0,0 +1,267 @@
+    }
+
+    /**
+     * return the subordinated unit handler
```
```suggestion
     * Return the subordinated unit handler
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetSystemBridgeHandler.java:
```diff
@@ -0,0 +1,267 @@
+    }
+
+    /**
+     * returns the list configuration of the units available for this system
```
```suggestion
     * Returns the list configuration of the units available for this system
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetSystemBridgeHandler.java:
```diff
@@ -0,0 +1,267 @@
+    }
+
+    /**
+     * return the id of this system
```
```suggestion
     * Return the id of this system
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetSystemBridgeHandler.java:
```diff
@@ -0,0 +1,267 @@
+    /**
+     * return the id of this system
+     * 
+     * @return
```
please add a short description

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetSystemBridgeHandler.java:
```diff
@@ -0,0 +1,267 @@
+    /**
+     * update the system state with the dto
+     * 
+     * @param systemState
```
please add a short description

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetSystemBridgeHandler.java:
```diff
@@ -0,0 +1,267 @@
+    }
+
+    /**
+     * update the system state with the dto
```
```suggestion
     * Update the system state with the dto
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetSystemBridgeHandler.java:
```diff
@@ -0,0 +1,267 @@
+    }
+
+    /**
+     * process the available fault messages
```
```suggestion
     * Process the available fault messages
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetSystemBridgeHandler.java:
```diff
@@ -0,0 +1,267 @@
+    /**
+     * process the available fault messages
+     * 
+     * @param faultMessages
```
please add a short description

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetSystemBridgeHandler.java:
```diff
@@ -0,0 +1,267 @@
+    /**
+     * Update the configuration of the system and the subordinated units
+     * 
+     * @param system
```
please add short descriptions

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetUnitThingHandler.java:
```diff
@@ -0,0 +1,268 @@
+    }
+
+    /**
+     * get the SubMenuEntryDTO for this unit
```
```suggestion
     * Get the SubMenuEntryDTO for this unit
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetUnitThingHandler.java:
```diff
@@ -0,0 +1,268 @@
+    }
+
+    /**
+     * get the MenuItemTabViewDTO for this unit
```
```suggestion
     * Get the MenuItemTabViewDTO for this unit
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetUnitThingHandler.java:
```diff
@@ -0,0 +1,268 @@
+    }
+
+    /**
+     * get the timstamp of the last valid call of updateValues
```
```suggestion
     * Get the timstamp of the last valid call of updateValues
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetUnitThingHandler.java:
```diff
@@ -0,0 +1,268 @@
+            paramDescriptionMap.put(param.valueId, param);
+            var channelId = new ChannelUID(thingId, param.parameterId.toString()); // "bindingId:type:thingId:1")
+            if (thing.getChannel(channelId) == null) {
+                logger.info("UnitThing: Create channel '{}'", channelId);
```
```suggestion
                logger.debug("UnitThing: Create channel '{}'", channelId);
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetUnitThingHandler.java:
```diff
@@ -0,0 +1,268 @@
+    }
+
+    /**
+     * update the values of the channels
```
```suggestion
     * Update the values of the channels
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetUnitThingHandler.java:
```diff
@@ -0,0 +1,268 @@
+    /**
+     * update the values of the channels
+     * 
+     * @param values
```
please add a short description

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetUnitThingHandler.java:
```diff
@@ -0,0 +1,268 @@
+    /**
+     * stores the state for the channel in stateCache and calls updateState of this Thing
+     * 
+     * @param channelId
```
please add short descriptions

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetUnitThingHandler.java:
```diff
@@ -0,0 +1,268 @@
+    }
+
+    /**
+     * stores the state for the channel in stateCache and calls updateState of this Thing
```
```suggestion
     * Stores the state for the channel in stateCache and calls updateState of this Thing
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,45 @@
+
+	<config-description uri="thing-type:wolfsmartset:account">
+		<parameter name="username" type="text" required="false">
+			<label>Wolf Smartset username</label>
```
```suggestion
			<label>Username</label>
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,45 @@
+			<label>Wolf Smartset username</label>
+		</parameter>
+		<parameter name="password" type="text" required="false">
+			<label>Wolf Smartset password</label>
```
```suggestion
			<label>Password</label>
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,45 @@
+			<context>password</context>
+		</parameter>
+		<parameter name="refreshIntervalConfiguration" type="integer" min="2" required="false" unit="m">
+			<label>Refresh Interval to update the configuration</label>
```
```suggestion
			<label>Structure Refresh Interval</label>
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,45 @@
+			<label>Wolf Smartset password</label>
+			<context>password</context>
+		</parameter>
+		<parameter name="refreshIntervalConfiguration" type="integer" min="2" required="false" unit="m">
```
This should be marked as "advanced"

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,45 @@
+			<description>Specifies the refresh interval in minutes</description>
+			<default>10</default>
+		</parameter>
+		<parameter name="refreshIntervalValues" type="integer" min="2" required="false" unit="s">
```
This should be marked as "advanced"

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,45 @@
+			<default>10</default>
+		</parameter>
+		<parameter name="refreshIntervalValues" type="integer" min="2" required="false" unit="s">
+			<label>Refresh Interval to update values and events</label>
```
```suggestion
			<label>States Refresh Interval</label>
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,45 @@
+			<default>15</default>
+		</parameter>
+		<parameter name="discoveryEnabled" type="boolean" required="false">
+			<label>Discovery Enabled</label>
```
```suggestion
			<label>Background Discovery</label>
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/handler/WolfSmartsetUnitThingHandler.java:
```diff
@@ -0,0 +1,268 @@
+        } else {
+            switch (parmeter.unit) {
+                case "bar":
+                    return new ChannelTypeUID("barometric-pressure");
```
This type does not seem to be defined in the thing-types.xml.

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,70 @@
+
+	<channel-type id="number">
+		<item-type>Number</item-type>
+		<label>number</label>
```
```suggestion
		<label>Number</label>
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,70 @@
+
+	<channel-type id="contact">
+		<item-type>Contact</item-type>
+		<label>contact</label>
```
```suggestion
		<label>Contact</label>
```

kaikreuzer reviewed (changes requested):
Many thanks for your patience, @BoBiene!
Please find my review comments below.

BoBiene commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/api/WolfSmartsetApi.java:
```diff
@@ -0,0 +1,609 @@
+            }
+
+        } catch (WolfSmartsetCloudException e) {
+            logger.info("Error logging on to Wolf Smartset ({}): {}", loginFailedCounter, e.getMessage());
```
We get here with bad password or a network issue - changed it to debug. 

AccountBridge is setting the state to offline in L200
````java
updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Authorization failed");
````

BoBiene commented on bundles/org.openhab.binding.wolfsmartset/src/main/java/org/openhab/binding/wolfsmartset/internal/api/WolfSmartsetApi.java:
```diff
@@ -0,0 +1,609 @@
+
+    void loginFailedCounterCheck() {
+        if (loginFailedCounter > 10) {
+            logger.warn("Repeated errors logging on to Wolf Smartset");
```
Thing status will update on its own. serviceToken is set to null, and the Bridge will go into "disconnected"

BoBiene commented:
@kaikreuzer can you please trigger a rebuild.

BoBiene commented:
@kaikreuzer anything missing from my side?

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,77 @@
+		<state readOnly="true" pattern="%.1f %unit%"/>
+	</channel-type>
+
+
```
```suggestion
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,77 @@
+		<config-description-ref uri="thing-type:wolfsmartset:system"/>
+	</bridge-type>
+
+
```
```suggestion
```

kaikreuzer commented on bundles/org.openhab.binding.wolfsmartset/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,77 @@
+		<config-description-ref uri="thing-type:wolfsmartset:unit"/>
+	</thing-type>
+
+
```
```suggestion
```

BoBiene commented on bundles/org.openhab.binding.wolfsmartset/README.md:
```diff
@@ -0,0 +1,148 @@
+
+The configuration via the interface is the recommended way:
+
+- System things (bridge) are discovered after Account thing (bridge) is set up
```
to get more prosa you need to find another volunteer ;)

BoBiene commented:
@kaikreuzer I hope I have now implemented everything so that this can be merged.

kaikreuzer reviewed (approved):
Lgtm, thank you!
